### ISSUE TITLE
Can we move imputation logic from records.py to taxdata repo?

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#688]
Recent discussion regarding issue #292 raises again the basic question Matt Jensen posed in issue #436: can imputations of variables derived from `puf.csv` data be moved out of `records.py` into the `taxdata` repository's `puf-cps-processing.py` script?

Can't we simply move the following functions from `records.py` to `puf-cps-processing.py`?

```
    def _impute_variables(self):
        """
        Impute variables in 2009 PUF Records data
        """
        self._cmbtp_itemizer = self._imputed_cmbtp_itemizer()
        self._cmbtp_standard = self.e62100 - self.e00100 + self.e00700
        # impute number of taxpayers
        self._txpyers[:] = np.where(
            np.logical_or(self.MARS == 2,
                          np.logical_or(self.MARS == 3, self.MARS == 6)),
            2., 1.)
        # impute number of extra standard deductions for aged
        self.numextra[:] = np.where(
            np.logical_or(self.AGERANGE >= 6,
                          np.logical_and(self.age >= 65, self.AGERANGE == 0)),
            self._txpyers, 0.)
        # impute the ratio of household head in total household income
        total = np.where(self.MARS == 2,
                         self.wage_head + self.wage_spouse, 0)
        earnings_split = np.where(total != 0,
                                  self.wage_head / total, 1.)
        one_minus_earnings_split = 1.0 - earnings_split
        self.e00200p[:] = earnings_split * self.e00200
        self.e00200s[:] = one_minus_earnings_split * self.e00200
        self.e00900p[:] = earnings_split * self.e00900
        self.e00900s[:] = one_minus_earnings_split * self.e00900
        self.e02100p[:] = earnings_split * self.e02100
        self.e02100s[:] = one_minus_earnings_split * self.e02100

    def _imputed_cmbtp_itemizer(self):
        """
        Private class method calls global function defined below.
        """
        return imputed_cmbtp_itemizer(self.e17500, self.e00100, self.e18400,
                                      self.e62100, self.e00700,
                                      self.p04470, self.e21040,
                                      self.e18500, self.e20800)

    # DO NOT MOVE def _extrapolate_2009_puf(self):

@vectorize([float64(float64, float64, float64,
                    float64, float64,
                    float64, float64,
                    float64, float64)])
def imputed_cmbtp_itemizer(e17500, e00100, e18400,
                           e62100, e00700,
                           p04470, e21040,
                           e18500, e20800):
    """
    Global function that calculates _cmbtp_itemizer values
    (uses vectorize decorator to speed up calculations with NumPy arrays)
    """
    # pylint: disable=too-many-arguments
    medical_limited = max(0., e17500 - max(0., e00100) * 0.075)
    medical_adjustment = min(medical_limited, 0.025 * max(0., e00100))
    state_adjustment = max(0, e18400)
    return (e62100 - medical_adjustment + e00700 + p04470 + e21040 -
            state_adjustment - e00100 - e18500 - e20800)
```

Moving these functions would require that the imputed variables be added to the `puf.csv` file, but that does not seem like a problem and we would not have to do the imputation calculations every time Tax-Calculator is run.  This change would also remove the hard-coded use of @vectorize from the `records.py` file, which is a desirable change as discussed recently in issue #292.

Have I missed something that makes this change more complicated than it seems?

@MattHJensen @feenberg @talumbau @Amy-Xu

**Comment by zrisher** [comment]
If the imputed variables are useful, wouldn't leaving this logic here allow non-standard-data users to provide less initial data while still making use of the imputed values?

**Comment by MattHJensen** [comment]
> Have I missed something that makes this change more complicated than it seems?

@martinholmer, I don't think you have missed anything pending @Amy-Xu's review of #694: (if @Amy-Xu agrees that the imputations rely on the 2009 puf then that would likely resolve @zrisher's query).

**Comment by Amy-Xu** [comment]
Moving these imputations to taxdata indeed seems to be the right thing to do. On Zach's concern, I feel we should probably find a place in Tax-Calculator to document all these imputation logic so Tax-Calculator users would be able to find it when needed.

**Comment by martinholmer** [comment]
Amy said:
> I feel we should probably find a place in Tax-Calculator to document all this imputation logic, so Tax-Calculator users would be able to find it when needed.

Seems sensible.  The actual code will be in the public `taxdata` repository's `puf-cps-processing.py` script, but yes we should add a comment to the Records class that points people to the `puf-cps-processing.py` script for the details on how the `puf.csv` data has been prepared.

@Amy-Xu @MattHJensen @zrisher

**Comment by martinholmer** [comment]
**Progress Report on Issue #688**

Recent enhancements involving the age variables (see pull request #701) have reduced the number of lines in the `Records.impute_variables()` method.

Pending pull request #721 further reduces by one the number of lines in `Records.impute_variables()`.

After #721 is merged, another pull request (analogous to #721) will eliminate the `_cmbtp_itemizer` line in the `Records.impute_variables()` method, moving the logic to the `functions.py` file.

After that, the only remaining logic in the `Records.impute_variables()` method will be the splitting of earnings between head (p) and spouse (s).  It seems as if that logic can be moved easily to the taxdata final-prep script.

@MattHJensen @feenberg @Amy-Xu @GoFroggyRun

**Comment by martinholmer** [comment]
Issue #688 has been resolved by the merge into master of pull request #754.
